feat(bigquery): add job_id_prefix to functions that produce bigquery loadjobs

dtran-im opened the pull request:
## Description of changes

I realized that my changes in https://github.com/ibis-project/ibis/pull/11233 and https://github.com/ibis-project/ibis/pull/11237 I completely focused on directly-created query jobs and completely overlooked functions which create bigquery load jobs and other queries, mainly in the process of uploading data.  Granted, the directly-created query jobs comprise the bulk of what I am concerned with, but it seemed proper to include other types of queries in this feature.

However, this has proved a bit stickier than anticipated.  The main situation I wasn't quite sure how to address was where one function call potentially creates several bigquery jobs, such as a job to truncate or drop a table if it already exists, a job to upload the data to a (temporary?) table in bq, and a job to select everything from that table and insert it into the target table.  I chose to address this with more specificity rather than less, creating kwargs such as `drop_job_id_prefix` and `insert_job_id_prefix`, and have attempted to be consistent about applying that, but I could be convinced that this is overkill and we should just pass along one `job_id_prefix` for any jobs stemming from a certain ibis function call.  (This would be ok for my purposes at least.)  

I believe I have added the kwargs to all relevant methods consistently, but it is worth a close review.  A notable exception is I did not add the kwargs to the `_run_pre_execute_hooks` call from `create_view`, since I assume that should not be creating a load job to upload data since it's creating a view, but that assumption could use verification.

It's notable that this required essentially copying the parent class' `insert`, `register_in_memory_tables`, and other methods in order to pass down a job_id_prefix variable.

In addition, since these methods do not return anything - there are no "results" from a drop table query that you would generally desire to return - I'm not immediately seeing a straightforward way to write unit tests for these.  But I have yet to explore some options.

## Issues closed

* Resolves #11229

dtran-im commented:
@cpcloud Do you have any insights on how I might make tests for this?  I fiddled a bit with patch & unittest.mock, but I haven't done much with those tools the past and any potential solution seemed exceedingly tortured.

dtran-im commented:
@cpcloud I added one test that I think may work, but I realized it's in the `/system/` tests which don't seem to get kicked off automatically.  Would you be able to kick those off?

cpcloud commented:
Thanks for the PR!

Is there a way we can centralize the API here?

I worry that every time there's a new need to customize some aspect of jobs, we'll have to modify N methods/functions' keyword arguments.

How are you producing the prefixes? Can we have some kind of callable that generates these that can be passed to `ibis.bigquery.connect` instead? Then, those callables would be internally passed around. That doesn't eliminate the problem of making sure to pass them wherever jobs are created, but it does central the customization of the IDs to a single call site.

dtran-im commented:
@cpcloud I've implemented the solution you suggested, though it feels a bit funny overwriting a class method inside a class method, if you think there's a better way of doing this let me know.  

The way I implemented it, we lose the ability to specify a particular job_id_prefix for a certain query in a more identifying way ("insert_items_into_table_query_" etc.), but I'm the one who requested this feature and just being able to generate and log a random uuid before the job kicks off suits my purposes.

dlstadther commented on ibis/backends/bigquery/__init__.py:
```diff
@@ -174,6 +175,78 @@ def _session_dataset(self):
+        destination: bq.TableReference | bq.Table | BqTableListItem | str,
+        **kwargs,
+    ) -> bq.LoadJob:
+        """Load a file into a BigQuery table with a job_id_prefix if specified."""
```
Should this docstring be updated? Currently, it's the exact same as `_client_load_table_from_file()`.

dlstadther commented on ibis/backends/bigquery/__init__.py:
```diff
@@ -1222,6 +1297,7 @@ def connect(
     auth_external_data: bool = False,
     auth_cache: str = "default",
     partition_column: str | None = "PARTITIONTIME",
+    generate_job_id_prefix: Callable[[], str | None] | None = None,
```
As i noted in https://github.com/ibis-project/ibis/issues/11230#issuecomment-2901159295 (which wasn't commented on), i think this `bigquery/__init__.py::connect()` function is legacy unused.

I'd like @cpcloud 's input, but i think it should be removed to avoid further contributor confusion.

dlstadther replied:
Thanks!

dlstadther commented on ibis/backends/bigquery/__init__.py:
```diff
@@ -174,6 +175,78 @@ def _session_dataset(self):
+    def _client_query(self, query: str, **kwargs) -> RowIterator:
+        """Run a query using the BigQuery client with a job_id_prefix if specified."""
+
+        job_id_prefix = kwargs.pop("job_id_prefix", self._generate_job_id_prefix())
```
In the current state, these `_client_*()` methods allow for the supply of `job_id_prefix` kwarg, but invocations of it never are given kwargs passthrough. Therefore, `job_id_prefix` will always use the `self._generate_job_id_prefix()` method.

For your use case (to log uuid job id before execution), how do you envision defining the `generate_job_id_prefix` callable such that you know the id its going to generate before the query executes? Would it be 1 id prefix per ibis client?

dtran-im replied:
Oh, yeah I should maybe modify it to just set `kwargs['job_id_prefix] = self._generate_job_id_prefix()`.

I was imagining an implementation something along the lines of:
```
def generate_job_id_prefix():
    prefix = uuid4().hex()
    logger.info(f"job_id_prefix: {prefix}")
    return prefix

ibis.bigquery.connect(..., generate_job_id_prefix=generate_job_id_prefix)
```
This way, before any query is made, we have the opportunity to log the job_id.  Then while the query is running, if we need to look into a particular query that's taking a long time, we can identify it.

dlstadther commented on ibis/backends/bigquery/__init__.py:
```diff
@@ -485,6 +583,10 @@ def do_connect(
 
         self.partition_column = partition_column
 
+        self._generate_job_id_prefix = (
+            generate_job_id_prefix or self._generate_job_id_prefix
+        )
```
Would it be less awkward to pass a callable that gets set as a class property? This could allow partial functions to be supplied?

It does seem awfully awkward to be overriding a static method on the object.

dtran-im replied:
It's perhaps a bit less awkward, though at least in my IDE the typing gets a little wonky.

dtran-im replied:
I changed it - how does it look now?

dlstadther commented:
@cpcloud Do you have additional feedback for this PR?

dtran-im commented on ibis/backends/bigquery/tests/system/test_client.py:
```diff
@@ -548,8 +553,59 @@ def test_window_with_count_distinct(tmp_table, expr, query):
+    con3.client.load_table_from_file = load_table_from_file
+
+    orig_query = con3.client.query
+    con3.client._query_num_calls = 0
```
I'll try that - I haven't used mocking much in unit tests before and wasn't quite sure how to achieve this

dtran-im replied:
I attempted to implement the mocker.spy and added a similar test for the `insert` method.  Let me know if there's anything more to do here.

dtran-im commented:
Ok, so I see a timeout, a memtable cleanup error for exasol, "feature not exposed in athena" errors, some 500 server errors, and some unknown operational errors among the failures.  I don't think these are relevant to my PR.

For the BQ failures though, I see a lot of tests failing due to `ValueError: Length mismatch: Expected axis has 1 elements, new values have 0 elements`.  TL;DR if my changes have broken things, I can't figure out how they possibly did.

This seems to stem from this statement in the `execute` function: `df.columns = schema.names`
Which the schema with zero elements comes from earlier in the `execute` function:
```
table_expr = expr.as_table()
schema = table_expr.schema() - ibis.schema({"_TABLE_SUFFIX": "string"})
```
I confirmed that this is the culprit, by imitating the test code and what `execute` runs, where `dtran.tmp` is a table that already exists in bq and contains data in the project specified:
```
>>> con = bigquery.connect(project_id="****")
>>> table = "dtran.tmp"
>>> limit = "LIMIT 10"
>>> expr = con.sql(f"SELECT * FROM {table} {limit}")
>>> expr.as_table()
SQLQueryResult
  query:
    SELECT * FROM dana_tran.tmp LIMIT 10
  schema:
    <empty schema>
>>> expr.as_table().schema()
ibis.Schema {
}
```
That's about as far as I've been able to get with my troubleshooting - I don't see any obvious changes to `.as_table()` or `.schema()`.  The bigquery backend method `get_schema` seems to work as intended:
```
>>> con.get_schema("tmp", database="dtran")
ibis.Schema {
  column_1  int64
  column_2  string
}
```

Is there a way to verify whether my changes are the culprit here?  (Do these tests pass on `main` currently?)  Thanks.

cpcloud commented on ibis/backends/bigquery/__init__.py:
```diff
@@ -670,16 +749,14 @@ def _make_session(self) -> tuple[str, str]:
 
     def _get_schema_using_query(self, query: str) -> sch.Schema:
-        job = self.client.query(
+        job = self._client_query(
```
The issue is here. `_get_schema_using_query` needs access to a `bq.QueryJob`, but `_client_query` always returns a `RowIterator`.

I'll fix this up!
